File: hasPath_DFS.py
"""
Input 1: a maze represented by a 2D array
0 0 1 0 0
0 0 0 0 0
0 0 0 1 0
1 1 0 1 1
0 0 0 0 0
Input 2: start coordinate (rowStart, colStart) = (0, 4)
Input 3: destination coordinate (rowDest, colDest) = (4, 4)
Output: true
Explanation: One possible way is : left -> down -> left -> down -> right -> down -> right.

Time complexity O(MN)

"""

class Solution:
    def hasPath(self,maze: List[List[int]],start: List[int],destination: List[int])->bool:
        if maze is None or len(maze)==0:
            return False 
        m=len(maze)
        n=len(maze[0])
        self.dirs=[[1,0],[0,-1],[0,1],[-1,0]]
        return self.dfs(maze,start,destination)
    
    def dfs(self,maze,start,destination):
        #base
        if(maze[start[0]][start[1]]==2):
            return False
        if(start[0]==destination[0] and start[1]==destination[1]):
                return True
        
        #logic
        maze[start[0]][start[1]]=2
        for dir in self.dirs:
            i=start[0]
            j=start[1]
            while(i>=0 and j>=0 and i<m and j<n and maze[i][j]!=1):
                    i=i+dir[0]
                    j=j+dir[1]
            i=i-dir[0]
            j=j-dir[1]
            if(self.dfs(maze,[i,j],destination)):
                return True
        return False
    
    

File: hasPath_DFS.cpp
// Time Complexity : O((m * n) * max(m, n)) 
// Space Complexity : O(m * n)
// Did this code successfully run on Leetcode : yes

// Three line explanation of solution in plain english
/* Perform DFS operation, change the direction only when the boundry or wall is hit.
 * To indicate visited cell, set its value to 2. 
 */

class Solution {
public:
    vector<vector<int>> dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    int m, n;  
    bool hasPath(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        
        if (maze.size() == 0)
        {
            return false;
        }
        
        m = maze.size();
        n = maze[0].size();
                
        return dfs(maze, start[0], start[1], destination);
    }
    
    bool dfs(vector<vector<int>>& maze, int& currRow, int& currCol, vector<int>& destination)
    {
        
        // base case
        
        if (currRow == destination[0] && currCol == destination[1])
        {
            return true;
        }
        
        if (maze[currRow][currCol] == 2)
        {
            return false;
        }

        // Logic
        
        maze[currRow][currCol] = 2;
        
        for (const auto& dir : dirs)
        {
            int i = currRow;
            int j = currCol;
                
            while (i < m && i >= 0 && j < n && j >= 0 && maze[i][j] != 1)
            {
                i += dir[0];
                j += dir[1];
            }
            
            i -= dir[0];
            j -= dir[1];
                                
            if (maze[i][j] == 0 && dfs(maze, i, j, destination))
            {
                return true;
            }
        }
        
        return false;
    }
};